op-batcher: exit process on criticial throttling RPC error

geoknee opened the pull request:
Confirmed manually that with these changes, the process is exited instead of just stopping the "work" of the batcher only.


Closes https://github.com/ethereum-optimism/optimism/issues/17768

geoknee committed:
op-batcher: thread CLI lifecycle context cancel function to driver

and wire up to fatal error condition for throttling loop

geoknee committed:
simplify wiring and abstract shutdownOnCriticalError method

geoknee committed:
introduce isCriticalThrottlingRPCError(err) helper

geoknee committed:
simplify impl

geoknee committed:
lint

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-crit

geoknee committed:
pass handler for closeApp in test environments

scharissis commented on op-batcher/batcher/driver.go:
```diff
@@ -671,6 +665,17 @@ func (l *BatchSubmitter) singleEndpointThrottler(wg *sync.WaitGroup, throttleSig
 }
 
+func isCriticalThrottlingRPCError(err error) bool {
+	var rpcErr rpc.Error
+	return errors.As(err, &rpcErr) && eth.ErrorCode(rpcErr.ErrorCode()).IsGenericRPCError()
+}
+
+func (l *BatchSubmitter) shutdownOnCriticalError(err error) {
+	l.Log.Error("Shutting down batcher on critical error", "err", err)
+	// Call closeApp to trigger process to exit (gracefully)
+	l.closeApp(err)
+}
```
Tests for this would be good; does it shut down? Is it graceful?

geoknee replied:
I tested this manually so far, and yes it does shut down successfully. I'm not totally sure we want to write a full end to end test for this, because we would need to spawn the batcher in a subprocess, attach it to an rpc endpoint returning `MethodNotFound`, and check on that process terminating. We don't have tests like this for sending an interrupt to the process, which should have the same effect. I'll see if I can add a little bit more unit tests to shore this up.

geoknee replied:
I extended an existing test here [a57e224](https://github.com/ethereum-optimism/optimism/pull/17924/commits/a57e2242cbbe1341ae03a0276db6d99460cee637). I'm pretty happy with this level of testing, it shows the "closeApp" is called under the right conditions. Having that `closeApp` actually cause the process to exit is arguably out of scope, since this is generic behavior from op-service that we rely on for multiple services.

scharissis commented on op-batcher/batcher/driver.go:
```diff
@@ -671,6 +665,17 @@ func (l *BatchSubmitter) singleEndpointThrottler(wg *sync.WaitGroup, throttleSig
 	}
 }
 
+func isCriticalThrottlingRPCError(err error) bool {
```
Tests for this would be good.

geoknee replied:
[59a967a](https://github.com/ethereum-optimism/optimism/pull/17924/commits/59a967a342b95f2fdbf747ff5be5cfb99dca5b08)

scharissis replied:
Nice! Short & Sweet.

geoknee committed:
add TestBatchSubmitter_CriticalError

geoknee committed:
slight improvement to test

geoknee committed:
extend TestBatchSubmitter_ThrottlingEndpoint to cover MethodNotFound case

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -99,6 +99,7 @@ type DriverSetup struct {
 // batches to L1 for availability.
 type BatchSubmitter struct {
 	DriverSetup
+	closeApp context.CancelCauseFunc
```
It looks to me like the `closeApp` naturally fits into the `DriverSetup`.

geoknee replied:
Did that here [e67dc90](https://github.com/ethereum-optimism/optimism/pull/17924/commits/e67dc90f439ed094045be1238cb599031e96db11)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -121,13 +122,14 @@ type BatchSubmitter struct {
 
 // NewBatchSubmitter initializes the BatchSubmitter driver from a preconfigured DriverSetup
-func NewBatchSubmitter(setup DriverSetup) *BatchSubmitter {
+func NewBatchSubmitter(setup DriverSetup, closeApp context.CancelCauseFunc) *BatchSubmitter {
```
...then we can also avoid passing it around like here.

geoknee replied:
Yeah good call [e67dc90](https://github.com/ethereum-optimism/optimism/pull/17924/commits/e67dc90f439ed094045be1238cb599031e96db11)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -671,6 +665,17 @@ func (l *BatchSubmitter) singleEndpointThrottler(wg *sync.WaitGroup, throttleSig
 }
 
+func isCriticalThrottlingRPCError(err error) bool {
+	var rpcErr rpc.Error
+	return errors.As(err, &rpcErr) && eth.ErrorCode(rpcErr.ErrorCode()).IsGenericRPCError()
+}
+
+func (l *BatchSubmitter) shutdownOnCriticalError(err error) {
+	l.Log.Error("Shutting down batcher on critical error", "err", err)
+	// Call closeApp to trigger process to exit (gracefully)
+	l.closeApp(err)
+}
```
If we want to allow using a `nil` closeApp, e.g. in tests, it may indeed make sense to add a nil check.

geoknee replied:
Added a nil check here [e67dc90](https://github.com/ethereum-optimism/optimism/pull/17924/commits/e67dc90f439ed094045be1238cb599031e96db11)

sebastianst commented on op-devstack/sysgo/l2_batcher.go:
```diff
@@ -115,9 +115,13 @@ func WithBatcher(batcherID stack.L2BatcherID, l1ELID stack.L1ELNodeID, l2CLID st
 		batcher, err := bss.BatcherServiceFromCLIConfig(
 			p.Ctx(), "0.0.1", batcherCLIConfig,
-			logger)
+			logger, closeAppFn)
```
So the devstack batcher wouldn't shut down?

geoknee replied:
Update that here [d6de699](https://github.com/ethereum-optimism/optimism/pull/17924/commits/d6de699b7b8ee224b4806eb72c9cc06807513c3a) so it should stop (there is no process to exit, but we cancel the context of the batcher service).

sebastianst replied:
that's better

sebastianst commented on op-batcher/batcher/service.go:
It may fit the pattern used in this file more if the `closeApp` is added as a field to the `BatcherService` struct and then read from there inside `initDriver` instead of passing it as an argument.

geoknee replied:
Fixed here [e67dc90](https://github.com/ethereum-optimism/optimism/pull/17924/commits/e67dc90f439ed094045be1238cb599031e96db11)

sebastianst commented on op-batcher/batcher/batch_submitter.go:
I think a slightly easier and more straight forward way of stopping the batcher is to call `BatcherService.Stop` when a critical error is found in the driver. It feels a bit wrong that we need to pass down a very high-level closer function down to the driver since the `BatcherService` already has a `Stop` function (that should be involved when the `closeApp` function is called that is passed down in the `batcher.Main` function). So we're still calling it, but more indirectly than necessary.

So instead, we can just pass down `BatcherService.Stop` to the driver when setting it up at `BatcherService.initDriver`.

Note that this way, we also don't need to create `closeApp` functions in the test setups that call `BatcherServiceFromCLIConfig`. And currently, the tests closer functions just make the tests fail, but still not stop the actual batcher service, if I see correctly.

Or am I missing something and you considered this slightly easier option?

geoknee replied:
On my first attempt at your suggestion, this doesn't seem to work and the batcher just hangs instead of fully shutting down. 

I think I went down the route of hooking up the driver to the very top level `closeApp` callback because it is the one that I know works (!) for fully shutting down the batcher. But I agree if we can wire things up without connecting something from the very outer / highest level to something deep inside the implementation, that would be better.

I think part of the problem is that the op-service app lifecycle context is not actually passed down to the `BatcherService`. So I had to wire "around" that gap. But this is something that could be fixed. I will investigate this further.

geoknee replied:
I think the driver shutdown logic might have a deadlock in it. But even removing all of the `wg.Done()` blocking calls, the process doesn't seem to actually exit when you call `BatcherService.Stop` (goroutines exit, but the process does not).

geoknee replied:
I dug into this pretty deeply, and I don't think we have any alternative to the main thrust of this PR (now tidied up a little bit), in terms of allowing the driver access to the entire app lifecycle context cancel function, unless we do a pretty major refactor of the batcher lifecycle itself. 

Ultimately, throttling is an optional routine, and if we want a condition there to close the entire app we either pass it the big red button directly, or we need to have various parts of the "parent" service block and wait on children service to return before themselves returning and causing a cascade which eventually quits the process. That's the only way I know where children don't need to know about their parents (main lifecycle cmd -> batcher service -> driver -> throttling loop -> individual throttler) but can still cause the app to quit. But that second architecture is pretty far outside of the op-service lifecycle paradigm and I do not think it is a good idea to try and modify that now.

sebastianst replied:
thanks for digging, agree...

sebastianst reviewed (commented):
I think there's an easier way (that actually also stops the batcher in tests).

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-crit

geoknee committed:
Embed closeApp in DriverSetup and BatcherService

Store closeApp in BatcherService and propagate it through DriverSetup so
the driver can access it without a separate parameter. Update
BatcherServiceFromCLIConfig/initFromCLIConfig signatures (closeApp moved
earlier), remove closeApp arg from NewBatchSubmitter, and guard
shutdownOnCriticalError against nil. Update call sites accordingly.

geoknee committed:
Propagate cancel cause to BatcherService

Use context.WithCancelCause and pass its cancel function when creating
the batcher so the service can be stopped with a cancel cause. Also
expose closeApp in DriverSetup and simplify a test by importing slices
and replacing manual loops with slices.Contains. Propagate cancel cause
to BatcherService

geoknee committed:
Use context.WithCancelCause for batcher

geoknee committed:
Update op-batcher/batcher/driver.go

Co-authored-by: almanax-ai[bot] <174396398+almanax-ai[bot]@users.noreply.github.com>

geoknee committed:
Use cancelable context and failure hook for batcher

Replace context.WithCancelCause with context.WithCancel and add a
closeAppFn that calls p.FailNow with the cause and cancels the batcher
context. Pass closeAppFn into BatcherServiceFromCLIConfig and import fmt
so tests fail immediately on critical batcher errors. Use cancelable
context and failure hook for batcher

Log the failure cause and cancel the batcher context on critical errors

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-crit

geoknee committed:
Use cancellable context for batcher

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -671,6 +668,19 @@ func (l *BatchSubmitter) singleEndpointThrottler(wg *sync.WaitGroup, throttleSig
+	if l.closeApp != nil {
+		// Call closeApp to trigger process to exit (gracefully) if l.closeApp is set.
+		l.closeApp(err)
+	}
```
maybe add a warn log in the `else` branch that there's no closeApp so possibly no shutdown happening?

geoknee committed:
Log shutdown error and warn if closeApp missing
